### PR TITLE
Refactor Activity launching to event-based approach

### PR DESCRIPTION
**Pull request by zsoltk**

### CODE REVIEW

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            )
+        }
+
+        events.getValue(id).accept(event)
+    }
+
+    protected fun Int.toInternalRequestCode() =
+        requestCodeRegistry.resolveRequestCode(this)
+
+    protected fun Identifiable.forgeExternalRequestCode(internalRequestCode: Int) =
```
Comment (ShikaSD) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt]: I see some inconsistency between `forgeCode`/`generateCode` methods.
Also the naming for `resolveRequestCode`/`resolveGroupId` is a bit confusing as well.
Should we align the naming?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    override fun events(client: Identifiable): Observable<T> {
+        val id = requestCodeRegistry.generateGroupId(client.id)
+        ensureSubject(id)
+
+        return events.getValue(id)
+            .doOnDispose { cleanup(id) }
+            .hide()
+    }
+
+    private fun ensureSubject(id: Int, onSubjectDidNotExist: (() -> Unit)? = null) {
```
Comment (ShikaSD) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt]: Rename `onSubjectDidNotExist` to `init`?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStream.kt [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package com.badoo.ribs.android
+
+import com.badoo.ribs.android.RequestCodeBasedEventStream.RequestCodeBasedEvent
+import com.badoo.ribs.core.Identifiable
+import io.reactivex.Observable
+
+interface RequestCodeBasedEventStream<T : RequestCodeBasedEvent> {
```
Comment (ShikaSD) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStream.kt]: How about renaming this and related classes to a bit shorter name?
`RequestCodeBasedEvent` sounds a bit too verbose.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
+package com.badoo.ribs.android
+
+import com.badoo.ribs.android.RequestCodeBasedEventStream.RequestCodeBasedEvent
+import com.badoo.ribs.core.Identifiable
+import com.badoo.ribs.core.requestcode.RequestCodeRegistry
+import com.jakewharton.rxrelay2.PublishRelay
+import com.jakewharton.rxrelay2.Relay
+import io.reactivex.Observable
+
+abstract class RequestCodeBasedEventStreamImpl<T : RequestCodeBasedEvent>(
```
Comment (ShikaSD) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt]: Not sure it makes sense to keep abstract `Impl` and the interface.
Right now you have bunch of other `Impl`s extending this one and interfaces extending base one.
Can we make it less scattered somehow?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt (hunk 2 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt]: It's used as a point to throw exceptions / log stuff, not to init

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStream.kt (hunk 3 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStream.kt]: Suggestions please

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt (hunk 4 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt]: I don't see a way right now to do that (interfaces extend each other, and so do implementations). If you have any concrete suggestions, please tell it.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt (hunk 2 above)
Comment (ShikaSD) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt]: Yeah, but from this method perspective it does not make sense

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt (hunk 1 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt]: Suggestions?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt (hunk 2 above)
Comment (ShikaSD) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt]: You can also make it return `Boolean` saying if the subject was created or not.

File: android/libraries/rib-android/src/main/java/com/uber/rib/core/RibActivity.kt [hunk 5]
```diff
@@ -4,22 +4,29 @@ import android.content.Intent
...
 import com.badoo.ribs.core.requestcode.RequestCodeRegistry
-import io.reactivex.Observable
 
-abstract class RibActivity : AppCompatActivity(),
-    ActivityStarter,
-    IntentCreator,
-    PermissionRequester {
+abstract class RibActivity : AppCompatActivity(), IntentCreator {
+
+    private val requestCodeRegistry =
```
Comment (lukaville) [inline android/libraries/rib-android/src/main/java/com/uber/rib/core/RibActivity.kt]: https://github.com/badoo/RIBs/pull/1#discussion_r258636323?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt (hunk 2 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt]: I started with boolean. The current one makes for a more expressive usage for me.

Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt]: + It's purely internal. Keeping it.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStream.kt (hunk 3 above)
Comment (Nublo) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStream.kt]: For example - `RequestCodeEvent`

File: android/libraries/rib-android/src/main/java/com/uber/rib/core/RibActivity.kt (hunk 5 above)
Comment (zsoltk) [reply android/libraries/rib-android/src/main/java/com/uber/rib/core/RibActivity.kt]: Fixed.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt (hunk 1 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt]: No suggestions - I'm leaving as is for now

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt (hunk 4 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStreamImpl.kt]: No suggestions - I'm leaving as is for now

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStream.kt (hunk 3 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/RequestCodeBasedEventStream.kt]: That's semantically not correct. I'll leave it for now as is.

**Pull request merged**